Dockerfile

```diff
@@ -29,10 +29,11 @@ RUN conda install -c bioconda -c conda-forge --file=requirements.txt \
     && conda clean --all -y
 
 COPY . .
-RUN python setup.py install
+RUN python setup.py install \
+    && rm -rf Autometa.egg-info/ build dist
 
-RUN hmmpress autometa/databases/markers/bacteria.single_copy.hmm \
-    && hmmpress autometa/databases/markers/archaea.single_copy.hmm
+RUN hmmpress -f autometa/databases/markers/bacteria.single_copy.hmm \
+    && hmmpress -f autometa/databases/markers/archaea.single_copy.hmm
```

Comment:
#132 This is being performed here. Perhaps you need to rebuild your docker image without the cache? This can be a problem sometimes.

```bash
docker build --no-cache . -t jason-c-kwan/autometa:nextflow
```

Reply:
# Autometa configuration within docker image

## Two new entrypoints for autometa configuration and database management

1. `autometa-config`
2. `autometa-update-databases`

### `autometa-config`

#132 will be resolved in this branch now with the use of the entrypoint `autometa-config` in the Dockerfile immediately after the HMMs are hmmpressed.

Note, this updates the autometa configuration (`autometa/config/default.config`) base databases directory

```bash
# autometa-config --section databases --option base --value autometa/databases
section : databases
option : base
value : autometa/databases
# Will now look for markers in autometa/databases/markers (where the HMMs are stored in the docker image)
```

This is updated via `autometa-config --section databases --option base --value autometa/databases`

The user similarly has the ability now to update any database and for autometa to look up this configuration later as the changes will be updated within the default.config file. Note, this can be done after installation from source or before and the user will easily have the ability to check their configuration at any stage by typing `autometa-config --print`

Since nextflow will be handling the checkpointing of steps in the pipeline, I've removed the code (`project.py`,` user.py`) that was a first attempt at managing this.

#### Usage

```bash
usage: autometa-config [-h] [--section {environ,databases,ncbi,markers}]
                       [--option OPTION] [--value VALUE] [--print]

Update Autometa configuration using provided arguments

optional arguments:
  -h, --help            show this help message and exit

Logging:
  --print               Print configuration without updating

Updating:
  --section {environ,databases,ncbi,markers}
                        config section to update
  --option OPTION       option in `--section` to update
  --value VALUE         Value to update `--option`
```

### `autometa-update-databases`

By default, with no arguments, will download/format databases into default
databases directory.

#### Usage

```bash
usage: autometa-update-databases [-h] [--config CONFIG] [--dryrun]
                                 [--update-all] [--update-markers]
                                 [--update-ncbi] [--check-dependencies]
                                 [--no-checksum] [--nproc NPROC] [--out OUT]

Main script to configure Autometa database dependencies.

optional arguments:
  -h, --help            show this help message and exit
  --config CONFIG       </path/to/input/database.config> (default: /home/evan/
                        miniconda3/envs/autometa/lib/python3.6/site-packages/A
                        utometa-2.0a0-py3.6.egg/autometa/config/default.config
                        )
  --dryrun              Log configuration actions but do not perform them.
                        (default: False)
  --update-all          Update all out-of-date databases. (default: False)
  --update-markers      Update out-of-date markers databases. (default: False)
  --update-ncbi         Update out-of-date ncbi databases. (default: False)
  --check-dependencies  Check database dependencies are satisfied. (default:
                        False)
  --no-checksum         Do not perform remote checksum comparisons to validate
                        databases are up-to-date. (default: False)
  --nproc NPROC         num. cpus to use for DB formatting. (default: 96)
  --out OUT             </path/to/output/database.config> (default: None)
```

Reply:
After installing rsync in the Docker image (see Issue #134), I still get these error messages when trying to run `autometa-update-databases`:

```bash
(base) root@d9acb2a58ecd:/test_data# autometa-update-databases --update-all --out database.config
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (ncbi,nodes)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (ncbi,merged)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (ncbi,nr)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (ncbi,accession2taxid)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (ncbi,names)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (markers,archaea_single_copy_cutoffs)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (markers,bacteria_single_copy_cutoffs)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (markers,archaea_single_copy)
[03/16/2021 05:52:57 PM DEBUG] autometa.config.databases: MISSING: (markers,bacteria_single_copy)
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/configparser.py", line 789, in get
    value = d[option]
  File "/opt/conda/lib/python3.8/collections/__init__.py", line 898, in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
  File "/opt/conda/lib/python3.8/collections/__init__.py", line 890, in __missing__
    raise KeyError(key)
KeyError: 'nodes'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/autometa-update-databases", line 33, in <module>
    sys.exit(load_entry_point('Autometa==2.0a0', 'console_scripts', 'autometa-update-databases')())
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 807, in main
    config = dbs.configure(section=section, no_checksum=args.no_checksum)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 702, in configure
    self.download_missing(section=section)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 574, in download_missing
    dispatcher[section](options)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 389, in download_ncbi_files
    ftp_fullpath = self.config.get("database_urls", option)
  File "/opt/conda/lib/python3.8/configparser.py", line 792, in get
    raise NoOptionError(option, section)
configparser.NoOptionError: No option 'nodes' in section: 'database_urls'
```

Reply:
OK, thanks for the update. But I am not getting a different error:

```bash
(base) root@08360f08fb7b:/test_data# autometa-update-databases --update-all --out database.config
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (ncbi,merged)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (ncbi,accession2taxid)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (ncbi,nodes)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (ncbi,nr)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (ncbi,names)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (markers,archaea_single_copy)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (markers,bacteria_single_copy_cutoffs)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (markers,archaea_single_copy_cutoffs)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: MISSING: (markers,bacteria_single_copy)
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: UPDATE: (ncbi,taxdump): autometa/databases/ncbi/taxdump.tar.gz
[03/16/2021 09:43:22 PM DEBUG] autometa.config.databases: starting taxdump download
[03/16/2021 09:43:27 PM DEBUG] autometa.common.utilities: Wrote autometa/databases/ncbi/taxdump.tar.gz checksum to autometa/databases/ncbi/taxdump.tar.gz.md5
[03/16/2021 09:43:29 PM DEBUG] autometa.config.databases: UPDATE: (ncbi,accession2taxid): autometa/databases/ncbi/prot.accession2taxid.gz
[03/16/2021 09:43:29 PM DEBUG] autometa.config.databases: starting accession2taxid download
Traceback (most recent call last):
  File "/opt/conda/bin/autometa-update-databases", line 33, in <module>
    sys.exit(load_entry_point('Autometa==2.0a0', 'console_scripts', 'autometa-update-databases')())
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 814, in main
    config = dbs.configure(section=section, no_checksum=args.no_checksum)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 709, in configure
    self.download_missing(section=section)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 581, in download_missing
    dispatcher[section](options)
  File "/opt/conda/lib/python3.8/site-packages/Autometa-2.0a0-py3.8.egg/autometa/config/databases.py", line 416, in download_ncbi_files
    subprocess.run(
  File "/opt/conda/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['rsync', '--quiet', '--archive', 'rsync://rsync.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.gz', 'autometa/databases/ncbi/prot.accession2taxid.gz']' returned non-zero exit status 10.
```

Looked up the rsync error and it looks like this occurs when rsync is not enabled on the remote server. Have NCBI removed this functionality? This doesn't seem quite right because it did already download taxdump.

Reply:
I'm having trouble recapitulating this error. Could you provide more information about your environment?